File: sdks/jvm/sdk/src/test/java/io/pact/janus/sdk/ScriptedEngine.java
package io.pact.janus.sdk;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import io.pact.janus.sdk.engine.Embedding;
import io.pact.janus.sdk.engine.FramePipe;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;

/**
 * An in-memory engine for the session-lifecycle and variant-iteration tests: answers each operation
 * from a script, and records every request frame it was sent. Its default script is a well-behaved
 * engine; a test overrides the operations it cares about.
 */
final class ScriptedEngine implements Embedding {
  static final ObjectMapper MAPPER = new ObjectMapper();

  /** One request the engine received. */
  record Call(String op, JsonNode body) {}

  final List<Call> calls = new ArrayList<>();
  int opened;
  int closed;
  private final Map<String, Function<JsonNode, String>> script = new HashMap<>();
  private int handles;
  private final List<String> served = new ArrayList<>();

  ScriptedEngine() {
    ok("engine/hello", "{\"protocol-version\":1,\"engine\":{\"name\":\"scripted\",\"version\":\"0\"},\"capabilities\":{}}");
    ok("consumer-session/create", "{\"session\":\"cs-1\"}");
    on("consumer-session/add-interaction", body -> "{\"ok\":{\"handle\":\"i-" + ++handles + "\"}}");
    ok("consumer-session/start-transport",
        "{\"endpoint\":{\"kind\":\"http\",\"host\":\"127.0.0.1\",\"port\":1,\"base-url\":\"http://127.0.0.1:1\"}}");
    variants("base");
    on("consumer-session/serve-variant", body -> {
      served.add(body.path("variant").asText());
      return "{\"ok\":{}}";
    });
    on("consumer-session/finalise", body -> "{\"ok\":{\"results\":[],\"contract\":"
        + "{\"$format\":\"janus-contract/1\",\"consumer\":{\"name\":\"c\"},\"provider\":{\"name\":\"p\"},\"interactions\":[]}}}");
  }

  /** Answers {@code op} with {@code ok}. */
  ScriptedEngine ok(String op, String okJson) {
    return on(op, body -> "{\"ok\":" + okJson + "}");
  }

  /** Answers {@code op} with an error document. */
  ScriptedEngine error(String op, String errorJson) {
    return on(op, body -> "{\"error\":" + errorJson + "}");
  }

  /**
   * Answers {@code op} with the frame members {@code answer} returns ({@code "ok":…} or
   * {@code "error":…}, without braces-wrapped type/id, which are added here).
   */
  ScriptedEngine on(String op, Function<JsonNode, String> answer) {
    script.put(op, answer);
    return this;
  }

  /** Makes {@code consumer-session/variants} select variants with these ids, labelled {@code v:<id>}. */
  ScriptedEngine variants(String... ids) {
    StringBuilder json = new StringBuilder("{\"variants\":[");
    for (int i = 0; i < ids.length; i++) {
      json.append(i == 0 ? "" : ",").append("{\"id\":\"").append(ids[i]).append("\",\"label\":\"v:").append(ids[i])
          .append("\",\"origin\":\"").append(i == 0 ? "base" : "covering").append("\",\"assignment\":[]}");
    }
    return ok("consumer-session/variants", json.append("],\"report\":{}}").toString());
  }

  List<String> ops() {
    return calls.stream().map(Call::op).toList();
  }

  List<String> served() {
    return served;
  }

  JsonNode body(String op) {
    return calls.stream().filter(c -> c.op().equals(op)).map(Call::body).findFirst().orElseThrow();
  }

  @Override
  public FramePipe open() {
    opened++;
    return new FramePipe() {
      @Override
      public byte[] call(byte[] requestFrame) {
        try {
          JsonNode frame = MAPPER.readTree(requestFrame);
          String op = frame.path("op").asText();
          calls.add(new Call(op, frame.path("body")));
          Function<JsonNode, String> answer = script.get(op);
          String members = answer == null
              ? "\"error\":{\"code\":\"operation-unsupported\",\"category\":\"protocol\",\"message\":\"no script\",\"details\":{\"op\":\"" + op + "\"}}"
              : strip(answer.apply(frame.path("body")));
          String response = "{\"type\":\"response\",\"id\":\"" + frame.path("id").asText() + "\"," + members + "}";
          return response.getBytes(StandardCharsets.UTF_8);
        } catch (Exception e) {
          throw new IllegalStateException(e);
        }
      }

      @Override
      public void close() {
        closed++;
      }
    };
  }

  private static String strip(String braced) {
    String t = braced.trim();
    return t.substring(1, t.length() - 1);
  }

  static ObjectNode json(String text) {
    try {
      return (ObjectNode) MAPPER.readTree(text);
    } catch (Exception e) {
      throw new IllegalArgumentException(e);
    }
  }
}

File: sdks/typescript/test/conformance/scripted-engine.ts
// The engine a `lifecycle` or `variants` case runs against: a frame pipe that records every
// operation the SDK sends and answers from the case's script (README §3). What these cases check
// is what the SDK sends and what it does with what comes back — not what an engine would decide.

import type { protocol } from "../../src/generated/index.js";
import type { FramePipe } from "../../src/index.js";
import type { ScriptedEngine } from "./cases.js";

export const CONSUMER = "web-app";
export const PROVIDER = "orders-api";

const DEFAULT_CONTRACT = {
  $format: "janus-contract/1",
  consumer: { name: CONSUMER },
  provider: { name: PROVIDER },
  interactions: [],
};

export class ScriptedPipe implements FramePipe {
  /** Every operation sent, in order. */
  readonly ops: string[] = [];
  /** Every request frame body, alongside `ops`. */
  readonly bodies: Record<string, unknown>[] = [];
  closed = false;
  readonly #script: ScriptedEngine;
  #handles = 0;

  constructor(script: ScriptedEngine = {}) {
    this.#script = script;
  }

  /** The first frame body sent for `op`. */
  bodyOf(op: string): Record<string, unknown> | undefined {
    const at = this.ops.indexOf(op);
    return at < 0 ? undefined : this.bodies[at];
  }

  async exchange(frame: protocol.RequestFrame): Promise<protocol.ResponseFrame> {
    this.ops.push(frame.op);
    this.bodies.push(frame.body);
    const scripted = this.#script.errors?.[frame.op];
    // `after` answers that many calls the default way first: the engine that dies mid-loop.
    const before = this.ops.filter((op) => op === frame.op).length - 1;
    if (scripted && before >= (scripted.after ?? 0)) {
      const { code, message, problems, details } = scripted as Record<string, unknown> & { code: string };
      return {
        type: "response",
        id: frame.id,
        error: {
          code,
          message: (message as string) ?? code,
          category: "request",
          details: { ...((details as Record<string, unknown>) ?? {}), ...(problems ? { problems } : {}) },
        } as protocol.EngineError,
      };
    }
    return { type: "response", id: frame.id, ok: this.#reply(frame.op) };
  }

  async close(): Promise<void> {
    this.closed = true;
  }

  #reply(op: string): Record<string, unknown> {
    switch (op) {
      case "engine/hello":
        return { "protocol-version": 1, engine: { name: "scripted", version: "0.0.0" }, capabilities: {} };
      case "consumer-session/create":
        return { session: "s-1" };
      case "consumer-session/add-interaction":
        return { handle: `i-${++this.#handles}` };
      case "consumer-session/start-transport":
        return { endpoint: { kind: "http", "base-url": "http://127.0.0.1:1" } };
      case "consumer-session/variants":
        return { variants: this.#script.variants ?? [{ id: "base", label: "base" }] };
      case "consumer-session/finalise":
        return {
          results: this.#script.results ?? this.#verifiedResults(),
          ...(this.#script["withhold-contract"] ? {} : { contract: this.#script.contract ?? DEFAULT_CONTRACT }),
        };
      default:
        return {};
    }
  }

  #verifiedResults(): Record<string, unknown>[] {
    return Array.from({ length: this.#handles }, (_, i) => ({ handle: `i-${i + 1}`, status: "verified" }));
  }
}
